File: learn-greek-easy-frontend/src/types/changelog.ts
/**
 * Changelog entry types for the Learn Greek Easy application.
 */

/** Tag classification for changelog entries */
export type ChangelogTag = 'new_feature' | 'bug_fix' | 'announcement';

/** Single changelog entry (public/localized) */
export interface ChangelogItem {
  id: string;
  title: string;
  content: string;
  tag: ChangelogTag;
  created_at: string;
  updated_at: string;
}

/** Paginated response for public changelog */
export interface ChangelogListResponse {
  total: number;
  page: number;
  page_size: number;
  items: ChangelogItem[];
}

/** Admin changelog entry with all language fields */
export interface ChangelogEntryAdmin {
  id: string;
  title_en: string;
  title_ru: string;
  content_en: string;
  content_ru: string;
  tag: ChangelogTag;
  created_at: string;
  updated_at: string;
}

/** Request body for creating a changelog entry */
export interface ChangelogCreateRequest {
  title_en: string;
  title_ru: string;
  content_en: string;
  content_ru: string;
  tag: ChangelogTag;
}

/** Request body for updating a changelog entry (all fields optional) */
export interface ChangelogUpdateRequest {
  title_en?: string;
  title_ru?: string;
  content_en?: string;
  content_ru?: string;
  tag?: ChangelogTag;
}

/** Paginated response for admin changelog list */
export interface ChangelogAdminListResponse {
  total: number;
  page: number;
  page_size: number;
  items: ChangelogEntryAdmin[];
}

/** Tag display configuration */
export const CHANGELOG_TAG_CONFIG: Record<
  ChangelogTag,
  {
    labelKey: string;
    colorClass: string;
  }
> = {
  new_feature: {
    labelKey: 'changelog:tag.newFeature',
    colorClass: 'bg-green-100 text-green-800 dark:bg-green-900 dark:text-green-300',
  },
  bug_fix: {
    labelKey: 'changelog:tag.bugFix',
    colorClass: 'bg-amber-100 text-amber-800 dark:bg-amber-900 dark:text-amber-300',
  },
  announcement: {
    labelKey: 'changelog:tag.announcement',
    colorClass: 'bg-blue-100 text-blue-800 dark:bg-blue-900 dark:text-blue-300',
  },
};

/** Languages supported for changelog content */
export type ChangelogLanguage = 'en' | 'ru';

/** Language options for admin form tabs */
export const CHANGELOG_LANGUAGES: readonly ChangelogLanguage[] = ['en', 'ru'];

/** Tag options for admin form select */
export const CHANGELOG_TAG_OPTIONS: readonly ChangelogTag[] = [
  'new_feature',
  'bug_fix',
  'announcement',
];

/** Get title field name for a specific language */
export function getTitleField(lang: ChangelogLanguage): keyof ChangelogEntryAdmin {
  return `title_${lang}` as keyof ChangelogEntryAdmin;
}

/** Get content field name for a specific language */
export function getContentField(lang: ChangelogLanguage): keyof ChangelogEntryAdmin {
  return `content_${lang}` as keyof ChangelogEntryAdmin;
}

File: learn-greek-easy-backend/src/schemas/changelog.py
"""Changelog entry Pydantic schemas for API request/response validation.

This module contains schemas for:
- Public changelog endpoint (localized content)
- Admin CRUD operations (all languages)
- Paginated listing responses
"""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from src.db.models import ChangelogTag

# ============================================================================
# Public Schemas (Localized - single language based on Accept-Language)
# ============================================================================


class ChangelogItemResponse(BaseModel):
    """Single changelog entry for public API (localized)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    content: str
    tag: ChangelogTag
    created_at: datetime
    updated_at: datetime


class ChangelogListResponse(BaseModel):
    """Paginated list of changelog entries (public)."""

    total: int = Field(..., ge=0)
    page: int = Field(..., ge=1)
    page_size: int = Field(..., ge=1, le=50)
    items: list[ChangelogItemResponse]


# ============================================================================
# Admin Schemas (All languages)
# ============================================================================


class ChangelogEntryCreate(BaseModel):
    """Schema for creating a changelog entry (admin)."""

    title_en: str = Field(..., min_length=1, max_length=500)
    title_ru: str = Field(..., min_length=1, max_length=500)
    content_en: str = Field(..., min_length=1)
    content_ru: str = Field(..., min_length=1)
    tag: ChangelogTag


class ChangelogEntryUpdate(BaseModel):
    """Schema for updating a changelog entry (admin). All fields optional."""

    title_en: Optional[str] = Field(None, min_length=1, max_length=500)
    title_ru: Optional[str] = Field(None, min_length=1, max_length=500)
    content_en: Optional[str] = Field(None, min_length=1)
    content_ru: Optional[str] = Field(None, min_length=1)
    tag: Optional[ChangelogTag] = None


class ChangelogEntryAdminResponse(BaseModel):
    """Full changelog entry for admin API (all languages)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title_en: str
    title_ru: str
    content_en: str
    content_ru: str
    tag: ChangelogTag
    created_at: datetime
    updated_at: datetime


class ChangelogAdminListResponse(BaseModel):
    """Paginated list of changelog entries (admin)."""

    total: int = Field(..., ge=0)
    page: int = Field(..., ge=1)
    page_size: int = Field(..., ge=1, le=50)
    items: list[ChangelogEntryAdminResponse]
